Sources/Commands/Utilities/SymbolGraphExtract.swift

```diff
@@ -48,12 +57,115 @@ public struct SymbolGraphExtract {
+                outputDirectory: outputDirectory.appending(component: "swift"),
+                verboseOutput: verboseOutput
+            )
+        } else if let clangTarget = target.underlyingTarget as? ClangTarget,
+                  case let .clang(buildDescription) = buildPlan.targetMap[target]
+        {
+            return try extractSymbolGraphFromClangTarget(
+                target: clangTarget,
+                buildDescription: buildDescription,
+                outputRedirection: outputRedirection,
+                outputDirectory: outputDirectory.appending(component: "clang"),
+                verboseOutput: verboseOutput
+            )
+        }
+    }
+    
+    func extractSymbolGraphFromClangTarget(
+        target: ClangTarget,
+        buildDescription: ClangTargetBuildDescription,
+        outputRedirection: Process.OutputRedirection,
+        outputDirectory: AbsolutePath,
+        verboseOutput: Bool
+    ) throws {
+        // Collect the set of public headers by filtering all of the target's headers to those
+        // inside the target's include directory.
+        var publicHeaders = Set(target.headers.filter(target.includeDir.isAncestor))
+        
+        guard !publicHeaders.isEmpty else {
+            return
+        }
+        
+        let umbrellaHeader: AbsolutePath?
+        switch target.moduleMapType {
+        case .umbrellaHeader(let umbrellaHeaderFile):
+            // The target is configured with an umbrella header – store its path separately
+            // and remove it from the general list of public headers.
+            umbrellaHeader = umbrellaHeaderFile
+            publicHeaders.remove(umbrellaHeaderFile)
+        case .umbrellaDirectory(_):
+            // TODO: Support symbol graph extraction for umbrella directory module map type
+            fallthrough
+        case .custom(_):
+            // TODO: Support symbol graph extraction for custom module map file type
+            fallthrough
+        case .none:
+            umbrellaHeader = nil
+        }
+        
+        try fileSystem.createDirectory(outputDirectory, recursive: true)
+        
+        // Construct the command line arguments for extracting the symbol graph
+        var commandLine = [
+            clangCompiler.pathString, "-extract-api",
+            "--product-name=\(target.c99name)",
+            "-o", outputDirectory.appending(component: "\(target.name).symbols.json").pathString
+        ]
+        
+        if verboseOutput {
+            commandLine += ["-v"]
+        }
+        
+        switch outputFormat {
+        case .json(let pretty):
+            if pretty {
+                // FIXME: Specify pretty print behavior to `-extract-api` if it's supported.
+                // commandLine += ["-pretty-print"]
+            }
+        }
+        
+        // FIXME: Does it matter for `-extract-api` usage if the target is C++ or not?
+        commandLine += try buildDescription.basicArguments(isCXX: nil)
```

Comment:
@daniel-grumberg should it matter if the target is C++ or not? The `basicArguments()` call adds some additional flags if so: https://github.com/apple/swift-package-manager/blob/fa91d29/Sources/Build/BuildPlan.swift#L417